zkevm-circuits/src/evm_circuit/execution/error_stack.rs

```diff
@@ -0,0 +1,352 @@
+
+        cb.opcode_stack_lookup(
+            opcode.expr(),
+            min_stack_pointer.expr(),
+            max_stack_pointer.expr(),
+        );
+        // Check whether current stack pointer is underflow or overflow
+
+        let is_overflow = LtGadget::construct(
+            cb,
+            cb.curr.state.stack_pointer.expr(),
+            min_stack_pointer.expr(),
+        );
+        let is_underflow = LtGadget::construct(
+            cb,
+            max_stack_pointer.expr(),
+            cb.curr.state.stack_pointer.expr(),
+        );
+
+        cb.require_boolean("is_overflow is bool", is_overflow.expr());
+        cb.require_boolean("is_underflow is bool", is_underflow.expr());
+
+        // constrain one of [is_underflow, is_overflow] must be true when stack error
+        // happens
+        cb.require_equal(
+            "is_underflow + is_overflow = 1",
+            is_underflow.expr() + is_overflow.expr(),
+            1.expr(),
+        );
+
+        // current call must be failed.
+        cb.call_context_lookup(false.expr(), None, CallContextFieldTag::IsSuccess, 0.expr());
+
+        // Go to EndTx only when is_root
+        let is_to_end_tx = cb.next.execution_state_selector([ExecutionState::EndTx]);
+        cb.require_equal(
+            "Go to EndTx only when is_root",
+            cb.curr.state.is_root.expr(),
+            is_to_end_tx,
+        );
+
+        // When it's a root call
+        cb.condition(cb.curr.state.is_root.expr(), |cb| {
+            // Do step state transition
+            cb.require_step_state_transition(StepStateTransition {
+                call_id: Same,
+                rw_counter: Delta(1.expr() + cb.curr.state.reversible_write_counter.expr()),
+                ..StepStateTransition::any()
+            });
+        });
+
+        // When it's an internal call, need to restore caller's state as finishing this
+        // call. Restore caller state to next StepState
+        let restore_context = cb.condition(1.expr() - cb.curr.state.is_root.expr(), |cb| {
+            RestoreContextGadget::construct(
+                cb,
+                0.expr(),
+                // rw_offset is handled in construct internally
+                0.expr(),
+                0.expr(),
+                0.expr(),
+                0.expr(),
+                0.expr(),
+            )
+        });
+
+        Self {
+            opcode,
+            min_stack_pointer,
+            max_stack_pointer,
+            is_overflow,
+            is_underflow,
+            restore_context,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _tx: &Transaction,
+        call: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        let opcode = step.opcode.unwrap();
+        let min_stack = opcode.stack_pair().0 as u64;
+        let max_stack = opcode.stack_pair().1 as u64;
+
```

Comment:
more rust style?

let (min_stack, max_stack) = opcode.stack_pair();
self.min_stack_pointer
            .assign(region, offset, Value::known(F::from(min_stack as u64)))?;
self.max_stack_pointer
    .assign(region, offset, Value::known(F::from(max_stack as u64)))?;

Reply:
good suggestion, I like it !

Reply:
updated in [3c9c55d](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/873/commits/3c9c55da2df49e8f6a236241b8d42fcfe6f4f465) thanks!